### PR TITLE
Metadata for user orgunits in analytics response [2.41-dhis2-15933]

### PR DESCRIPTION
**Pull request by d-bernat**
please see https://dhis2.atlassian.net/browse/DHIS2-15933

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/AnalyticsMetaDataKey.java [hunk 1]
```diff
@@ -36,11 +36,16 @@ public enum AnalyticsMetaDataKey {
   ITEMS("items"),
   DIMENSIONS("dimensions"),
   PAGER("pager"),
+  ORG_UNITS("organisationUnits"),
   ORG_UNIT_HIERARCHY("ouHierarchy"),
   ORG_UNIT_NAME_HIERARCHY("ouNameHierarchy"),
-  ORG_UNIT_ANCESTORS("ouAncestors");
+  ORG_UNIT_ANCESTORS("ouAncestors"),
+  USER_ORGUNIT("USER_ORGUNIT"),
```
Comment (gnespolino) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/AnalyticsMetaDataKey.java:43]: does it need to have a different case respect the other keys ? Couldn't it be `userOrgUnit` ?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/AnalyticsMetaDataKey.java [hunk 2]
```diff
@@ -36,11 +36,16 @@ public enum AnalyticsMetaDataKey {
...
 
-  private String key;
+  USER_ORGUNIT_CHILDREN("USER_ORGUNIT_CHILDREN"),
+  USER_ORGUNIT_GRANDCHILDREN("USER_ORGUNIT_GRANDCHILDREN");
```
Comment (gnespolino) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/AnalyticsMetaDataKey.java:46]: same as previous comment

**Review by gnespolino — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java [hunk 3]
```diff
@@ -383,20 +387,28 @@ protected void addMetadata(
...
       } else {
         optionItems.addAll(getItemOptionsAsFilter(params.getItemOptions(), params.getItems()));
       }
 
+      Map<String, Object> items = new HashMap<>();
+      AnalyticsOrganisationUnitUtils.getUserOrganisationUnitsUidList(
+              currentUserService.getCurrentUser(), params.getUserOrganisationUnitsCriteria())
+          .forEach(items::putAll);
+
+      if (params.isComingFromQuery()) {
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java:400]: Could we concentrate all the logic of the if/else related to
```
params.isComingFromQuery()
```
on the same block? ie:

```
      if (params.isComingFromQuery()) {
        items.putAll(getMetadataItems(params, periodKeywords, optionItems, grid));
        metadata.put(
            DIMENSIONS.getKey(), getDimensionItems(params, Optional.of(optionsPresentInGrid)));
      } else {
        items.putAll(getMetadataItems(params));
        metadata.put(DIMENSIONS.getKey(), getDimensionItems(params, empty()));
      }

      metadata.put(ITEMS.getKey(), items);
```
Or is it going to have a side affect?

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/AnalyticsOrganisationUnitUtils.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import org.hisp.dhis.analytics.AnalyticsMetaDataKey;
+import org.hisp.dhis.common.BaseIdentifiableObject;
+import org.hisp.dhis.user.User;
+
+/** Utilities for organisation unit criteria of outcoming analytic response */
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/AnalyticsOrganisationUnitUtils.java:45]:
```suggestion
/** Utilities for organisation unit criteria of outcoming analytics response. */
```

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/AnalyticsOrganisationUnitUtils.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class AnalyticsOrganisationUnitUtils {
+  /**
+   * Retrieve collection of all uids of organisation units belongs to current user and present in
+   * response grid.
+   *
+   * @param currentUser the {@link org.hisp.dhis.user.CurrentUser}.
+   * @return intersection of requested user organisation units and all units in response grid
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/AnalyticsOrganisationUnitUtils.java:53]:
```suggestion
   * @return intersection of requested user organisation units and all units in response grid.
```

**Review by maikelarabori — CHANGES_REQUESTED** [review]
Added a few suggestions 👍 
Also, this same logic should be applied to the TEI/Cross Program API (I believe). The UI requests are built using the same components and are read on the same page.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/OrganisationUnitCriteriaUtils.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+   */
+  public static String getAnalyticsQueryCriteria(AggregateAnalyticsQueryCriteria criteria) {
+    return hasDimensions(criteria.getDimension())
+        ? criteria.getDimension().stream()
+            .filter(d -> d.contains(ORGUNIT_DIM_ID))
+            .collect(Collectors.joining(","))
+        : StringUtils.EMPTY;
+  }
+
+  private static boolean hasDimensions(Set<String> dimensions) {
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/OrganisationUnitCriteriaUtils.java:118]: done

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/OrganisationUnitCriteriaUtils.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            c ->
+                c.equalsIgnoreCase(AnalyticsMetaDataKey.USER_ORGUNIT.getKey())
+                    || c.equalsIgnoreCase(AnalyticsMetaDataKey.USER_ORGUNIT_CHILDREN.getKey())
+                    || c.equalsIgnoreCase(AnalyticsMetaDataKey.USER_ORGUNIT_GRANDCHILDREN.getKey()))
+        .map(AnalyticsMetaDataKey::valueOf)
+        .toList();
+  }
+
+  /**
+   * Transform criteria in the string
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/OrganisationUnitCriteriaUtils.java]: done

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java (hunk 3 above)
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java:400]: the items will change the place with dimensions, I thing this should not be a problem, should I change it?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/OrganisationUnitCriteriaUtils.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+import org.hisp.dhis.common.EnrollmentAnalyticsQueryCriteria;
+import org.hisp.dhis.common.EventsAnalyticsQueryCriteria;
+
+/** Utilities for organisation unit criteria of incoming analytic request */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class OrganisationUnitCriteriaUtils {
+  /**
+   * Translate user organisation unit criteria in analytics keys (MetaData)
+   *
+   * @param userOrganisationUnitsCriteria {@link String}.
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/OrganisationUnitCriteriaUtils.java:52]: done

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/OrganisationUnitCriteriaUtils.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        .toList();
+  }
+
+  /**
+   * Transform criteria in the string
+   *
+   * @param criteria {@link EnrollmentAnalyticsQueryCriteria}.
+   * @return string of criteria
+   */
+  public static String getAnalyticsQueryCriteria(EnrollmentAnalyticsQueryCriteria criteria) {
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/OrganisationUnitCriteriaUtils.java:84]: done

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/OrganisationUnitCriteriaUtils.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/OrganisationUnitCriteriaUtils.java]: done

**Pull request merged**